hbw/tasks/inference.py

```diff
@@ -0,0 +1,332 @@
+
+# import luigi
+import law
+
+from columnflow.tasks.framework.base import Requirements, RESOLVE_DEFAULT
+from columnflow.tasks.framework.parameters import SettingsParameter
+from columnflow.tasks.framework.mixins import (
+    CalibratorsMixin, SelectorStepsMixin, ProducersMixin, MLModelsMixin, InferenceModelMixin,
+)
+from columnflow.tasks.framework.remote import RemoteWorkflow
+from columnflow.tasks.cms.inference import CreateDatacards
+from columnflow.util import dev_sandbox, maybe_import
+
+array = maybe_import("array")
+
+
+def get_hist_name(cat_name: str, proc_name: str, syst_name: str | None = None) -> str:
+    hist_name = f"{cat_name}/{proc_name}"
+    if syst_name:
+        hist_name += f"__{syst_name}"
+    return hist_name
+
+
+def get_cat_proc_syst_names(root_file):
+    cat_names = set()
+    proc_names = set()
+    syst_names = set()
+    for key in root_file.keys():
+        # remove ";1" appendix
+        key = key.split(";")[0]
+
+        key = key.split("/")
+        # first part of the key is the category
+        cat_names.add(key[0])
+
+        # second part is the hist name: <proc_name>__<syst_name>
+        if len(key) == 2:
+            hist_name = key[1].split("__")
+
+            proc_name = hist_name[0]
+            proc_names.add(proc_name)
+
+            if len(hist_name) == 2:
+                syst_name = hist_name[1]
+                syst_names.add(syst_name)
+
+    return cat_names, proc_names, syst_names
+
+
+def get_rebin_values(hist, N_bins_final: int = 10):
+    """
+    Function that determines how to rebin a hist to *N_bins_final* bins such that
+    the resulting histogram is flat
+    """
+    N_bins_input = hist.GetNbinsX()
+
+    # determine events per bin the final histogram should have
+    events_per_bin = hist.Integral() / N_bins_final
+    print(f" ============ {round(events_per_bin, 3)} events per bin")
+
+    # bookkeeping number of bins and number of events
+    bin_count = 1
+    N_events = 0
+    rebin_values = [hist.GetBinLowEdge(1)]
+
+    # starting loop at 1 to exclude underflow
+    # ending at N_bins_input + 1 excludes overflow
+    for i in range(1, N_bins_input + 1):
+        if bin_count == N_bins_final:
+            # break as soon as N-1 bin edges have been determined --> final bin is x_max
+            break
+
+        N_events += hist.GetBinContent(i)
+        if i % 100 == 0:
+            print(f" ========== Bin {i} of {N_bins_input}, {N_events} events")
+        if N_events >= events_per_bin * bin_count:
+            # when *N_events* surpasses threshold, append the corresponding bin edge and count
+            print(f" ++++++++++ Append bin edge {bin_count} of {N_bins_final} at edge {hist.GetBinLowEdge(i)}")
+            rebin_values.append(hist.GetBinLowEdge(i + 1))
+            bin_count += 1
+
+    # final bin is x_max
+    x_max = hist.GetBinLowEdge(N_bins_input + 1)
+    rebin_values.append(x_max)
+    print(f"final bin edges: {rebin_values}")
+    return rebin_values
```

Comment:
A general question: The histogram is supposed to be flat. Does that mean it is mandatory, that all bins have the same number of events or is it possible, that they are different by a small amount (i.e. the histogram is only approxamitely flat)?

Reply:
It's just approximately flat, since I am only rebinning existing histograms. I am simply counting the number of entries of the histogram and as soon as the number of entries is higher than some threshold, I am appending another bin edge.

If the result deviates too far from a flat distribution, we could simply increase the number of bins from our input histograms (which will get computationally expensive at some point) or we would need to determine our bin edges before histogramming (which will require lots of memory)